docs/runtime/ir-caching.md

```diff
@@ -0,0 +1,209 @@
+problem. Instead, we solve it using a preprocessing step:
+
+- We can modify `BindingsMap` and its child types to be able to contain an
+  unlinked module pointer
+  `case class ModulePointer(namespace: String, qualifiedName: List[String])` in
+  place of a `Module`.
+- As the `MetadataStorage` type that holds the `BindingsMap` is mutable it can
+  be updated in place without having to reassemble the entire IR graph.
+- Hence, we can traverse all the nodes in the `ir.preorder` that have metadata
+  consisting of either the `BindingsMap` or `ResolvedName` types (provided by
+  the following passes: `BindingAnalysis`, `MethodDefinitions`,
+  `UppercaseNames`, `VectorLiterals`, `Patterns`), and perform a replacement.
+
+Having done this, we have broken any links that the IR may hold between modules,
+and can serialise each module individually.
+
+## Storing the IR
+
+The serialized IR needs to be stored in a location that is tied to the library
+that it serializes. Despite this, we _also_ want to be able to ship cached IR
+with libraries. This leads to a two pronged solution where we check two
+locations for the cache.
+
+1. **With the Library:** As libraries can have a hidden `.enso` directory, we
+   can use a path within that for caching. This should be
+   `$package/.enso/cache/ir/enso-$version/`, and can be got at by extending the
+   `pkg` library to be aware of the cache directories.
+2. **Globally:** As some library locations may not be writable, we need to have
+   a global out-of-line cache that is used if the first one is not writable.
+   This is located under `$ENSO_DATA` (whose location can be obtained from the
+   `RuntimeDistributionManager`), and is located under the path
+   `$ENSO_DATA/cache/ir/$hash/enso-$version/`, where `$hash` is the `SHA3-224`
+   hash of the tuple `(namespace, library_name, version)`, where
+   `version = SemVer | "local"`.
+
+In each location, the IR is stored with the following assumptions:
+
+- The IR file is located in a directory modelled after its module path, followed
+  by a file named after the module itself with the extension `.ir` (e.g. the IR
+  for `Standard.Base.Data.Vector.enso` is stored in
+  `Standard/Base/Data/Vector.ir`).
+- The [metadata](#metadata-format) file is located in a directory modelled after
+  its module path, followed by a file named after the module itself with the
+  extension `.meta` (e.g. the metadata for `Standard.Base.Data.Vector.enso` is
+  stored in `Standard/Base/Data/Vector.meta`).
+
+Storage of the IR only takes place iff the intended location for that IR is
+_empty_.
+
+### Metadata Format
+
+The metadata is used for integrity checking of the cached IR to prevent loading
+corrupted or out of date data from the cache. It is _exceedingly_ important to
+make the cache decoding forwards-compatible. It should be able to ignore unknown
+fields when loading, allowing us to update the logic in the future if needed.
```

Comment:
Actually is it that important? After all, we store the cache in separate directories for each engine version, so actually it a newer engine version will not be able to access caches from other versions.